File: src/services/tauri/insights.ts
import { invoke } from '@tauri-apps/api/core';

export interface DayActivity {
  date: string;
  count: number;
  size: number;
}

export interface ToolCallStats {
  tool_name: string;
  count: number;
}

export interface TokenStats {
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_read_tokens: number;
  reasoning_tokens?: number;
  total_tokens: number;
}

export interface HeatmapData {
  agent: string;
  data: DayActivity[];
  max_count: number;
  total_files: number;
  total_size: number;
  tool_calls: ToolCallStats[];
  token_stats: TokenStats;
  /** Unique model names seen, sorted by frequency */
  models: string[];
}

export interface AgentHeatmaps {
  claude: HeatmapData | null;
  codex: HeatmapData | null;
  gemini: HeatmapData | null;
}

export interface InsightFilters {
  range?: string;      // "day" | "week" | "month" | "year" | "all"
  cwd?: string;
  session_id?: string;
  agent?: string;      // "Claude" | "Codex" | "Gemini"
}

export interface FilterOptions {
  cwds: string[];
  session_ids: string[];
}

export async function getAgentHeatmaps(filters: InsightFilters): Promise<AgentHeatmaps> {
  return invoke<AgentHeatmaps>('get_agent_heatmaps', {
    range: filters.range ?? null,
    cwd: filters.cwd ?? null,
    sessionId: filters.session_id ?? null,
    agent: filters.agent ?? null,
  });
}

export async function getInsightFilterOptions(): Promise<FilterOptions> {
  return invoke<FilterOptions>('get_insight_filter_options');
}

export interface RankItem {
  key: string;
  sessions: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  agents: string[];
}

export interface Rankings {
  by_cwd: RankItem[];
  by_session: RankItem[];
}

export async function getInsightRankings(filters: InsightFilters): Promise<Rankings> {
  return invoke<Rankings>('get_insight_rankings', {
    range: filters.range ?? null,
    cwd: filters.cwd ?? null,
    sessionId: filters.session_id ?? null,
    agent: filters.agent ?? null,
  });
}

File: src-tauri/src/commands/insights.rs
pub use codexia_shared::insights::{AgentHeatmaps, FilterOptions, Rankings};

#[tauri::command]
pub async fn get_agent_heatmaps(
    range: Option<String>,
    cwd: Option<String>,
    session_id: Option<String>,
    agent: Option<String>,
) -> Result<AgentHeatmaps, String> {
    codexia_shared::insights::get_agent_heatmaps(range, cwd, session_id, agent).await
}

#[tauri::command]
pub async fn get_insight_rankings(
    range: Option<String>,
    cwd: Option<String>,
    session_id: Option<String>,
    agent: Option<String>,
) -> Result<Rankings, String> {
    codexia_shared::insights::get_insight_rankings(range, cwd, session_id, agent).await
}

#[tauri::command]
pub async fn get_insight_filter_options() -> Result<FilterOptions, String> {
    codexia_shared::insights::get_insight_filter_options().await
}
